File: code/shared/schemas/rhizome_snapshot.py
"""Schema RhizomeSnapshot."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field, field_validator

from .base import BaseSchema, parse_utc_zulu
from .enums import Mode


class SnapshotSensors(BaseModel):
    soil_moisture_a_pct: float = Field(ge=0, le=100)
    soil_moisture_b_pct: float = Field(ge=0, le=100)
    tank_level_pct: float = Field(ge=0, le=100)
    flow_rate_lpm: float = Field(ge=0)
    last_reading_at: datetime

    @field_validator("last_reading_at", mode="before")
    @classmethod
    def validate_last_reading_at(cls, value: Any) -> datetime:
        return parse_utc_zulu(value)


class SnapshotVision(BaseModel):
    last_capture_at: datetime | None = None
    parcel_a_status: str | None = None
    parcel_b_status: str | None = None
    visual_notes: str | None = Field(default=None, max_length=500)

    @field_validator("last_capture_at", mode="before")
    @classmethod
    def validate_last_capture_at(cls, value: Any) -> datetime | None:
        if value is None:
            return None
        return parse_utc_zulu(value)


class SnapshotHealth(BaseModel):
    cpu_temp_c: float
    cpu_load_pct: float = Field(ge=0, le=100)
    ram_used_gb: float = Field(ge=0)
    ram_total_gb: float = Field(gt=0)
    esp32_heartbeat_ok: bool
    last_esp32_ack_at: datetime | None = None

    @field_validator("cpu_temp_c")
    @classmethod
    def validate_cpu_temp_c(cls, value: float) -> float:
        if not -20 <= value <= 120:
            raise ValueError("cpu_temp_c fuera de rango razonable")
        return value

    @field_validator("last_esp32_ack_at", mode="before")
    @classmethod
    def validate_last_esp32_ack_at(cls, value: Any) -> datetime | None:
        if value is None:
            return None
        return parse_utc_zulu(value)


class RhizomeSnapshot(BaseSchema):
    snapshot_id: str = Field(min_length=1)
    mode: Mode
    active_policy_id: str = Field(min_length=1)
    active_policy_expires_at: datetime
    sensors: SnapshotSensors
    vision: SnapshotVision
    health: SnapshotHealth
    last_decision_id: str | None = None
    pending_contradictions: list[str] = Field(default_factory=list)

    @field_validator("active_policy_expires_at", mode="before")
    @classmethod
    def validate_active_policy_expires_at(cls, value: Any) -> datetime:
        return parse_utc_zulu(value)

File: code/pollen/app/src/main/kotlin/net/sprout/pollen/schemas/RhizomeSnapshot.kt
package net.sprout.pollen.schemas

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class RhizomeSnapshot(
    @SerialName("schema_version") val schemaVersion: String = "1.0",
    @SerialName("created_at") val createdAt: String,
    @SerialName("origin_node_id") val originNodeId: String,
    @SerialName("signature") val signature: String? = null,
    
    @SerialName("snapshot_id") val snapshotId: String,
    @SerialName("mode") val mode: Mode,
    
    @SerialName("active_policy_id") val activePolicyId: String,
    @SerialName("active_policy_expires_at") val activePolicyExpiresAt: String,
    
    @SerialName("sensors") val sensors: Sensors,
    @SerialName("vision") val vision: Vision,
    @SerialName("health") val health: Health,
    
    @SerialName("last_decision_id") val lastDecisionId: String? = null,
    @SerialName("pending_contradictions") val pendingContradictions: List<String> = emptyList()
) {
    @Serializable
    data class Sensors(
        @SerialName("soil_moisture_a_pct") val soilMoistureAPct: Float,
        @SerialName("soil_moisture_b_pct") val soilMoistureBPct: Float,
        @SerialName("tank_level_pct") val tankLevelPct: Float,
        @SerialName("flow_rate_lpm") val flowRateLpm: Float,
        @SerialName("last_reading_at") val lastReadingAt: String
    )

    @Serializable
    data class Vision(
        @SerialName("last_capture_at") val lastCaptureAt: String? = null,
        @SerialName("parcel_a_status") val parcelAStatus: String? = null,
        @SerialName("parcel_b_status") val parcelBStatus: String? = null,
        @SerialName("visual_notes") val visualNotes: String? = null
    )

    @Serializable
    data class Health(
        @SerialName("cpu_temp_c") val cpuTempC: Float,
        @SerialName("cpu_load_pct") val cpuLoadPct: Float,
        @SerialName("ram_used_gb") val ramUsedGb: Float,
        @SerialName("ram_total_gb") val ramTotalGb: Float,
        @SerialName("esp32_heartbeat_ok") val esp32HeartbeatOk: Boolean,
        @SerialName("last_esp32_ack_at") val lastEsp32AckAt: String? = null
    )
}
